Android Remote Debug: Godot Editor remains stuck if one-click deploy device architecture doesn't match the device

### Tested versions

v4.2.1.stable.official [b09f793f5]

### System information

Godot v4.2.1.stable - macOS 14.2.0 - Vulkan (Mobile) - integrated Apple M1 - Apple M1 (8 Threads)

### Issue description

Godot does not yet check whether the architecture fits. Progress bar remains stuck at 66%, it is not possible to press "Cancel". The application "does not respond" and must be terminated via the task manager.

![Bildschirmfoto 2023-12-18 um 02 23 22](https://github.com/godotengine/godot/assets/41921395/4ef46d11-6390-41b9-a100-624faad5dc27)


### Steps to reproduce

In the export dialog, select the Android CPU architecture that the device does **not** support.

![Bildschirmfoto 2023-12-18 um 01 23 06](https://github.com/godotengine/godot/assets/41921395/9521b870-ae96-45f8-9ae9-89620a13654f)
![Bildschirmfoto 2023-12-18 um 01 23 35](https://github.com/godotengine/godot/assets/41921395/065d9012-0d9d-41c4-9dc1-e23704414aea)


### Minimal reproduction project (MRP)

N/A

`NoClassDefFoundError`, maybe because my tablet is looking for the classes under '/lib/armeabi-v7a'

<details><summary>logcat</summary>

```
12-18 01:18:42.959 18629 18629 I zygote  : Rejecting re-init on previously-failed class java.lang.Class<org.godotengine.godot.Godot$onInitRenderView$2>: java
.lang.NoClassDefFoundError: Failed resolution of: Landroid/view/WindowInsetsAnimation$Callback;
12-18 01:18:42.959 18629 18629 I zygote  :   at void org.godotengine.godot.GodotFragment.onCreate(android.os.Bundle) (GodotFragment.java:181)
12-18 01:18:42.959 18629 18629 I zygote  :   at void androidx.fragment.app.Fragment.performCreate(android.os.Bundle) (Fragment.java:2949)
12-18 01:18:42.959 18629 18629 I zygote  :   at void androidx.fragment.app.FragmentStateManager.create() (FragmentStateManager.java:475)
12-18 01:18:42.959 18629 18629 I zygote  :   at void androidx.fragment.app.FragmentStateManager.moveToExpectedState() (FragmentStateManager.java:278)
12-18 01:18:42.959 18629 18629 I zygote  :   at void androidx.fragment.app.FragmentManager.executeOpsTogether(java.util.ArrayList, java.util.ArrayList, int, 
int) (FragmentManager.java:2189)
12-18 01:18:42.959 18629 18629 I zygote  :   at void androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(java.util.ArrayList, java.util
.ArrayList) (FragmentManager.java:2100)
12-18 01:18:42.959 18629 18629 I zygote  :   at void androidx.fragment.app.FragmentManager.execSingleAction(androidx.fragment.app.FragmentManager$OpGenerator
, boolean) (FragmentManager.java:1971)
12-18 01:18:42.959 18629 18629 I zygote  :   at void androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss() (BackStackRecord.java:311)
12-18 01:18:42.959 18629 18629 I zygote  :   at void org.godotengine.godot.GodotActivity.onCreate(android.os.Bundle) (GodotActivity.kt:84)
12-18 01:18:42.959 18629 18629 I zygote  :   at void com.godot.game.GodotApp.onCreate(android.os.Bundle) (GodotApp.java:45)
12-18 01:18:42.959 18629 18629 I zygote  :   at void android.app.Activity.performCreate(android.os.Bundle, android.os.PersistableBundle) (Activity.java:7258)
12-18 01:18:42.959 18629 18629 I zygote  :   at void android.app.Activity.performCreate(android.os.Bundle) (Activity.java:7249)
12-18 01:18:42.959 18629 18629 I zygote  :   at void android.app.Instrumentation.callActivityOnCreate(android.app.Activity, android.os.Bundle) (Instrumentati
on.java:1222)
12-18 01:18:42.959 18629 18629 I zygote  :   at android.app.Activity android.app.ActivityThread.performLaunchActivity(android.app.ActivityThread$ActivityClie
ntRecord, android.content.Intent) (ActivityThread.java:2927)
12-18 01:18:42.959 18629 18629 I zygote  :   at void android.app.ActivityThread.handleLaunchActivity(android.app.ActivityThread$ActivityClientRecord, android
.content.Intent, java.lang.String) (ActivityThread.java:3059)
12-18 01:18:42.959 18629 18629 I zygote  :   at void android.app.ActivityThread.-wrap11(android.app.ActivityThread, android.app.ActivityThread$ActivityClient
Record, android.content.Intent, java.lang.String) (ActivityThread.java:-1)
12-18 01:18:42.959 18629 18629 I zygote  :   at void android.app.ActivityThread$H.handleMessage(android.os.Message) (ActivityThread.java:1724)
12-18 01:18:42.959 18629 18629 I zygote  :   at void android.os.Handler.dispatchMessage(android.os.Message) (Handler.java:106)
12-18 01:18:42.959 18629 18629 I zygote  :   at void android.os.Looper.loop() (Looper.java:164)
12-18 01:18:42.959 18629 18629 I zygote  :   at void android.app.ActivityThread.main(java.lang.String[]) (ActivityThread.java:7000)
12-18 01:18:42.959 18629 18629 I zygote  :   at java.lang.Object java.lang.reflect.Method.invoke(java.lang.Object, java.lang.Object[]) (Method.java:-2)
12-18 01:18:42.959 18629 18629 I zygote  :   at void com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run() (RuntimeInit.java:441)
12-18 01:18:42.959 18629 18629 I zygote  :   at void com.android.internal.os.ZygoteInit.main(java.lang.String[]) (ZygoteInit.java:1408)
12-18 01:18:42.959 18629 18629 I zygote  : Caused by: java.lang.ClassNotFoundException: Didn't find class "android.view.WindowInsetsAnimation$Callback" on pa
th: DexPathList[[zip file "/data/app/com.example.mrcollisionexample-_FsJlWjfZxJBa6MPnPbCYw==/base.apk"],nativeLibraryDirectories=[/data/app/com.example.mrcollisionexample-_FsJlWjfZxJBa6MPnPbCYw==/lib/arm, /data/app/com.example.mrcollisionexample-_FsJlWjfZxJBa6MPnPbCYw==/base.apk!/lib/armeabi-v7a, /system/lib, /system/vendor/lib]]
```
</details>

--- a/editor/editor_run_native.cpp
+++ b/editor/editor_run_native.cpp
@@ -79,6 +79,11 @@ void EditorRunNative::_notification(int p_what) {
 	}
 }
 
+void EditorRunNative::_confirm_run_native() {
+	run_confirmed = true;
+	resume_run_native();
+}
+
 Error EditorRunNative::start_run_native(int p_id) {
 	if (p_id < 0) {
 		return OK;
@@ -86,9 +91,9 @@ Error EditorRunNative::start_run_native(int p_id) {
 
 	int platform = p_id / 10000;
 	int idx = p_id % 10000;
+	resume_id = p_id;
 
 	if (!EditorNode::get_singleton()->ensure_main_scene(true)) {
-		resume_id = p_id;
 		return OK;
 	}
 
@@ -109,6 +114,22 @@ Error EditorRunNative::start_run_native(int p_id) {
 		EditorNode::get_singleton()->show_warning(TTR("No runnable export preset found for this platform.\nPlease add a runnable preset in the Export menu or define an existing preset as runnable."));
 		return ERR_UNAVAILABLE;
 	}
+
+	String architecture = eep->get_device_architecture(idx);
+	if (!run_confirmed && !architecture.is_empty()) {
+		String preset_arch = "architectures/" + architecture;
+		bool is_arch_enabled = preset->get(preset_arch);
+
+		if (!is_arch_enabled) {
+			String warning_message = vformat(TTR("Warning: The CPU architecture '%s' is not active in your export preset.\n\n"), Variant(architecture));
+			warning_message += TTR("Run 'Remote Debug' anyway?");
+
+			run_native_confirm->set_text(warning_message);
+			run_native_confirm->popup_centered();
+			return OK;
+		}
+	}
+	run_confirmed = false;
 
 	emit_signal(SNAME("native_run"), preset);
 
@@ -173,6 +194,10 @@ EditorRunNative::EditorRunNative() {
 
 	add_child(result_dialog);
 	result_dialog->hide();
+
+	run_native_confirm = memnew(ConfirmationDialog);
+	add_child(run_native_confirm);
+	run_native_confirm->connect("confirmed", callable_mp(this, &EditorRunNative::_confirm_run_native));
 
 	set_process(true);
 }

--- a/editor/editor_run_native.h
+++ b/editor/editor_run_native.h
@@ -41,11 +41,15 @@ class EditorRunNative : public HBoxContainer {
 
 	RichTextLabel *result_dialog_log = nullptr;
 	AcceptDialog *result_dialog = nullptr;
+	ConfirmationDialog *run_native_confirm = nullptr;
+	bool run_confirmed = false;
 
 	MenuButton *remote_debug = nullptr;
 	bool first = true;
 
 	int resume_id = -1;
+
+	void _confirm_run_native();
 
 protected:
 	static void _bind_methods();

--- a/editor/export/editor_export_platform.h
+++ b/editor/export/editor_export_platform.h
@@ -227,6 +227,7 @@ public:
 	virtual Ref<ImageTexture> get_option_icon(int p_index) const;
 	virtual String get_option_label(int p_device) const { return ""; }
 	virtual String get_option_tooltip(int p_device) const { return ""; }
+	virtual String get_device_architecture(int p_device) const { return ""; }
 
 	enum DebugFlags {
 		DEBUG_FLAG_DUMB_CLIENT = 1,

--- a/platform/android/export/export_plugin.cpp
+++ b/platform/android/export/export_plugin.cpp
@@ -379,7 +379,8 @@ void EditorExportPlatformAndroid::_check_for_changes_poll_thread(void *ud) {
 							} else if (p.begins_with("ro.build.version.sdk=")) {
 								d.api_level = p.get_slice("=", 1).to_int();
 							} else if (p.begins_with("ro.product.cpu.abi=")) {
-								d.description += "CPU: " + p.get_slice("=", 1).strip_edges() + "\n";
+								d.architecture = p.get_slice("=", 1).strip_edges();
+								d.description += "CPU: " + d.architecture + "\n";
 							} else if (p.begins_with("ro.product.manufacturer=")) {
 								d.description += "Manufacturer: " + p.get_slice("=", 1).strip_edges() + "\n";
 							} else if (p.begins_with("ro.board.platform=")) {
@@ -1990,6 +1991,12 @@ String EditorExportPlatformAndroid::get_option_tooltip(int p_index) const {
 		s = devices[p_index].name + "\n\n" + s;
 	}
 	return s;
+}
+
+String EditorExportPlatformAndroid::get_device_architecture(int p_index) const {
+	ERR_FAIL_INDEX_V(p_index, devices.size(), "");
+	MutexLock lock(device_lock);
+	return devices[p_index].architecture;
 }
 
 Error EditorExportPlatformAndroid::run(const Ref<EditorExportPreset> &p_preset, int p_device, int p_debug_flags) {

--- a/platform/android/export/export_plugin.h
+++ b/platform/android/export/export_plugin.h
@@ -76,6 +76,7 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 		String name;
 		String description;
 		int api_level = 0;
+		String architecture;
 	};
 
 	struct APKExportData {
@@ -220,6 +221,8 @@ public:
 	virtual String get_option_label(int p_index) const override;
 
 	virtual String get_option_tooltip(int p_index) const override;
+
+	virtual String get_device_architecture(int p_index) const override;
 
 	virtual Error run(const Ref<EditorExportPreset> &p_preset, int p_device, int p_debug_flags) override;
 

